pintracker/stateless/stateless.go

```diff
@@ -0,0 +1,490 @@
+// were updated or have errors. Cids in error states can be recovered
+// with Recover().
+// An error is returned if we are unable to contact the IPFS daemon.
+func (spt *Tracker) SyncAll() ([]api.PinInfo, error) {
+	// get ipfs status for all
+	localpis, err := spt.localStatus(false)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	for _, p := range spt.optracker.FilterOps(optracker.OperationPin, optracker.PhaseError) {
+		if _, ok := localpis[p.Cid().String()]; ok {
+			spt.optracker.Clean(p)
+		}
+	}
+
+	for _, p := range spt.optracker.FilterOps(optracker.OperationUnpin, optracker.PhaseError) {
+		if _, ok := localpis[p.Cid().String()]; ok {
+			spt.optracker.Clean(p)
+		}
+	}
+
+	return spt.getErrorsAll(), nil
+}
+
+// Sync returns the updated local status for the given Cid.
+func (spt *Tracker) Sync(c *cid.Cid) (api.PinInfo, error) {
+	return spt.Status(c), nil
+}
+
+// RecoverAll attempts to recover all items tracked by this peer.
+func (spt *Tracker) RecoverAll() ([]api.PinInfo, error) {
+	statuses := spt.StatusAll()
+	resp := make([]api.PinInfo, 0)
+	for _, st := range statuses {
+		r, err := spt.Recover(st.Cid)
+		if err != nil {
+			return resp, err
+		}
+		resp = append(resp, r)
+	}
+	return resp, nil
+}
+
+// Recover will re-track or re-untrack a Cid in error state,
+// possibly retriggering an IPFS pinning operation and returning
+// only when it is done.
+func (spt *Tracker) Recover(c *cid.Cid) (api.PinInfo, error) {
+	logger.Infof("Attempting to recover %s", c)
+	pInfo, ok := spt.getError(c)
+	if !ok {
+		return spt.Status(c), nil
+	}
+
+	var err error
+	switch pInfo.Status {
+	case api.TrackerStatusPinError:
+		err = spt.enqueue(api.PinCid(c), optracker.OperationPin)
+	case api.TrackerStatusUnpinError:
+		err = spt.enqueue(api.PinCid(c), optracker.OperationUnpin)
+	}
+	if err != nil {
+		return spt.Status(c), err
+	}
+
+	pi, ok := spt.optracker.GetExists(c)
+	if !ok {
+		return spt.Status(c), nil
+	}
+
+	return pi, nil
+}
+
+func (spt *Tracker) ipfsStatusAll() (map[string]api.PinInfo, error) {
+	var ipsMap map[string]api.IPFSPinStatus
+	err := spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"IPFSPinLs",
+		"recursive",
+		&ipsMap,
+	)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+	pins := make(map[string]api.PinInfo, 0)
+	for cidstr, ips := range ipsMap {
+		c, err := cid.Decode(cidstr)
+		if err != nil {
+			logger.Error(err)
+			continue
+		}
+		p := api.PinInfo{
+			Cid:    c,
+			Peer:   spt.peerID,
+			Status: ips.ToTrackerStatus(),
+		}
+		pins[cidstr] = p
+	}
+	return pins, nil
+}
+
+// localStatus returns a joint set of consensusState and ipfsStatus
+// marking pins which should be remote and leaving any ipfs pins that
+// aren't in the consensusState out.
+func (spt *Tracker) localStatus(incRemote bool) (map[string]api.PinInfo, error) {
+	pininfos := make(map[string]api.PinInfo)
+
+	// get shared state
+	var csps []api.PinSerial
```

Comment:
I'd appreciate longer, more explicit, variable names. We have `cpsp`, `cspis`.. -> `statePinsSerial`, `statePins` etc. for example. or make sure all variables have comments about what they are storing.

Reply:
✅